File: libraries/cpp/src/binary.cpp
#include "ifcx/binary.h"

#include <cstring>
#include <stdexcept>

namespace ifcx {

static constexpr char IFCX_MAGIC[] = "IFCX";
static constexpr size_t IFCX_MAGIC_SIZE = 4;

std::vector<uint8_t> IfcxbEncoder::encode(const IfcxDocument& /*doc*/) {
    throw std::runtime_error("IfcxbEncoder::encode is not yet implemented");
}

IfcxDocument IfcxbDecoder::decode(const std::vector<uint8_t>& data) {
    // Validate magic bytes
    if (data.size() < IFCX_MAGIC_SIZE) {
        throw std::runtime_error("Invalid IFCXB data: too short");
    }

    if (std::memcmp(data.data(), IFCX_MAGIC, IFCX_MAGIC_SIZE) != 0) {
        throw std::runtime_error("Invalid IFCXB data: missing 'IFCX' magic header");
    }

    throw std::runtime_error("IfcxbDecoder::decode is not yet implemented");
}

} // namespace ifcx

File: libraries/rust/src/binary.rs
//! IFCXB binary format encoder/decoder.

use crate::{IfcxDocument, IfcxError};

const MAGIC: &[u8; 4] = b"IFCX";

pub struct IfcxbEncoder;
pub struct IfcxbDecoder;

impl IfcxbEncoder {
    /// Encode document to IFCXB binary.
    pub fn encode(_doc: &IfcxDocument) -> Result<Vec<u8>, IfcxError> {
        // TODO: Implement IFCXB encoding
        Err(IfcxError::NotImplemented("IFCXB encoding".to_string()))
    }
}

impl IfcxbDecoder {
    /// Decode IFCXB binary to document.
    pub fn decode(data: &[u8]) -> Result<IfcxDocument, IfcxError> {
        if data.len() < 16 {
            return Err(IfcxError::InvalidBinary("file too short".to_string()));
        }
        if &data[0..4] != MAGIC {
            return Err(IfcxError::InvalidBinary("bad magic bytes".to_string()));
        }
        // TODO: Implement IFCXB decoding
        Err(IfcxError::NotImplemented("IFCXB decoding".to_string()))
    }
}
